File: pca2.py
# %%
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from latex_io import format_matrix

plt.rcParams.update(
    {
        "font.size": 14,
        "axes.titlesize": 16,
        "axes.labelsize": 14,
        "xtick.labelsize": 12,
        "ytick.labelsize": 12,
        "legend.fontsize": 12,
        "figure.titlesize": 18,
    }
)
plt.rcParams["figure.figsize"] = (8, 5)

# %%
# import dataset
df = pd.read_excel("data/Assignment_Dataset_2.xlsx", sheet_name=0)


# %%
# do pca for df, but ignore gender and accomodation_class
features = ["DSRI", "GMI", "AQI", "SGI", "DEPI", "SGAI", "ACCR", "LEVI"]
X = df[features].to_numpy()
mean = np.mean(X, axis=0)
std = np.std(X, axis=0)
X_std = (X - mean) / std  # standardize data
varicance = np.var(X_std)
cov = np.cov(X_std.T)  # covariance matrix
print("cov matrix: ")
print(format_matrix(cov, 2))

# %%
# Eigenvalues / Eigenvectors
eigenvalues, eigenvectors = np.linalg.eigh(cov)
idx = np.argsort(eigenvalues)[::-1]
eigenvalues = eigenvalues[idx]
eigenvectors = eigenvectors[:, idx]
print("Eigenvalues:")
print(eigenvalues)
print("Sum eigenvalues: ", sum(eigenvalues))

# Projection onto principal components
PCs = X_std @ eigenvectors
print("Shape of PCs:", PCs.shape)

explained_variance = eigenvalues / np.sum(eigenvalues)

print("Explained variance:")
for i, var in enumerate(explained_variance):
    print(f"PC{i+1}: {var:.3f}")
labels = ["PC1", "PC2", "PC3", "PC4", "PC5", "PC6", "PC7", "PC8"]
# plt.figure(figsize=(8, 5))
for i in range(5):
    plt.hist(PCs[:, i], bins=20, alpha=0.5, density=True, label=labels[i])

plt.xlabel("Principal Component Value")
plt.ylabel("Density")
plt.legend()
plt.savefig("pictures/pca2.pdf", bbox_inches="tight")
plt.show()

# %%
# Scree
# plt.figure(figsize=(6, 4))
plt.plot(range(1, len(explained_variance) + 1), explained_variance, marker="o")
plt.xticks(range(1, len(labels) + 1), labels)
plt.xlabel("Principal Component")
plt.ylabel("Explained Variance Ratio")
plt.grid(True)
plt.savefig("pictures/screeplot2.pdf", bbox_inches="tight")
plt.show()

# %%
loadings = pd.DataFrame(eigenvectors, index=features, columns=labels)
print(loadings)
plt.figure(figsize=(7, 4))
plt.imshow(loadings, cmap="coolwarm", aspect="auto")
plt.xticks(range(len(labels)), labels)
plt.yticks(range(len(features)), features)
plt.colorbar(label="Loading")
plt.xlabel("Principal Components")
plt.ylabel("Original Features")
plt.savefig("pictures/loading_pca2.pdf", bbox_inches="tight")
plt.show()


# %%
# plt.figure(figsize=(7, 7))
for i, feature in enumerate(features):
    plt.arrow(
        0,
        0,
        eigenvectors[i, 0],
        eigenvectors[i, 1],
        head_width=0.03,
        length_includes_head=True,
    )
    plt.text(eigenvectors[i, 0] * 1.1, eigenvectors[i, 1] * 1.1, feature, fontsize=11)

plt.axhline(0, color="gray", linewidth=0.5)
plt.axvline(0, color="gray", linewidth=0.5)
plt.xlabel("PC1")
plt.ylabel("PC2")
plt.axis("equal")
plt.grid(True)
plt.show()


# %%

colors = {"Yes": "tab:blue", "No": "tab:red"}

# plt.figure(figsize=(7, 6))

for decision in colors:
    mask = df["Manipulater"] == decision

    plt.scatter(PCs[mask, 0], PCs[mask, 1], label=decision, alpha=0.7)  # PC1  # PC2

plt.xlabel(f"PC1 ({explained_variance[0]*100:.1f}% Varianz)")
plt.ylabel(f"PC2 ({explained_variance[1]*100:.1f}% Varianz)")
plt.legend()
plt.grid(True)
plt.savefig("pictures/pc1vspc2_2.pdf", bbox_inches="tight")
plt.show()


# %%

File: pca2.R
# =====================================================
# PCA Analysis - Dataset 2
# =====================================================

library(readxl)

# -----------------------------------------------------
# Load data
# -----------------------------------------------------

df <- read_excel("data/Assignment_Dataset_2.xlsx", sheet = 1)

features <- c(
  "DSRI", "GMI", "AQI", "SGI",
  "DEPI", "SGAI", "ACCR", "LEVI"
)

X <- df[, features]

# -----------------------------------------------------
# Standardize data
# -----------------------------------------------------

X_std <- scale(X)

cov_matrix <- cov(X_std)

cat("Covariance matrix:\n")
print(round(cov_matrix, 2))

# -----------------------------------------------------
# PCA
# -----------------------------------------------------

pca <- prcomp(
  X,
  center = TRUE,
  scale. = TRUE
)

eigenvalues <- pca$sdev^2
eigenvectors <- pca$rotation
PCs <- pca$x

cat("\nEigenvalues:\n")
print(eigenvalues)

cat("\nSum of eigenvalues:\n")
print(sum(eigenvalues))

cat("\nShape of PCs:\n")
print(dim(PCs))

explained_variance <- eigenvalues / sum(eigenvalues)

cat("\nExplained variance:\n")

for(i in seq_along(explained_variance)){
  cat(sprintf("PC%d: %.3f\n",
              i,
              explained_variance[i]))
}

labels <- paste0("PC",1:8)

# -----------------------------------------------------
# Histograms of the first five principal components
# -----------------------------------------------------

hist(
  PCs[,1],
  probability = TRUE,
  breaks = 20,
  col = rgb(0,0,1,0.3),
  xlab = "Principal Component Value",
  main = ""
)

for(i in 2:5){
  
  hist(
    PCs[,i],
    probability = TRUE,
    breaks = 20,
    col = rgb(i/5,0,1-i/5,0.3),
    add = TRUE
  )
  
}

legend(
  "topright",
  legend = labels[1:5],
  fill = sapply(1:5,
                function(i) rgb(i/5,0,1-i/5,0.3))
)

# -----------------------------------------------------
# Scree plot
# -----------------------------------------------------

plot(
  explained_variance,
  type = "b",
  xaxt = "n",
  xlab = "Principal Component",
  ylab = "Explained Variance Ratio"
)

axis(
  1,
  at = 1:8,
  labels = labels
)

grid()

# -----------------------------------------------------
# Loadings
# -----------------------------------------------------

loadings <- eigenvectors

cat("\nLoadings:\n")
print(round(loadings,3))

image(
  t(loadings[nrow(loadings):1, ]),
  axes = FALSE,
  col = heat.colors(100)
)

axis(
  1,
  at = seq(0,1,length.out = 8),
  labels = labels
)

axis(
  2,
  at = seq(0,1,length.out = 8),
  labels = rev(features)
)

# -----------------------------------------------------
# Loading plot (PC1 vs PC2)
# -----------------------------------------------------

plot(
  c(-1,1),
  c(-1,1),
  type = "n",
  asp = 1,
  xlab = "PC1",
  ylab = "PC2"
)

abline(h = 0, col = "gray")
abline(v = 0, col = "gray")

for(i in 1:length(features)){
  
  arrows(
    0,
    0,
    eigenvectors[i,1],
    eigenvectors[i,2],
    length = 0.1
  )
  
  text(
    eigenvectors[i,1]*1.1,
    eigenvectors[i,2]*1.1,
    labels = features[i]
  )
  
}

grid()

# -----------------------------------------------------
# Scores: PC1 vs PC2
# -----------------------------------------------------

colors <- c(
  "Yes" = "blue",
  "No" = "red"
)

plot(
  PCs[,1],
  PCs[,2],
  col = colors[df$Manipulater],
  pch = 16,
  xlab = sprintf(
    "PC1 (%.1f%% Variance)",
    explained_variance[1]*100
  ),
  ylab = sprintf(
    "PC2 (%.1f%% Variance)",
    explained_variance[2]*100
  )
)

legend(
  "topright",
  legend = names(colors),
  col = colors,
  pch = 16
)

grid()